source/file.c

```diff
@@ -34,54 +44,91 @@ FILE *aws_fopen(const char *file_path, const char *mode) {
     return file;
 }
 
-int aws_byte_buf_init_from_file(struct aws_byte_buf *out_buf, struct aws_allocator *alloc, const char *filename) {
+static int s_byte_buf_init_from_file(
+    struct aws_byte_buf *out_buf,
+    struct aws_allocator *alloc,
+    const char *filename,
+    bool use_file_size_as_hint,
+    size_t size_hint) {
```

Comment:
Debatable / Code Clarity: There are two different ways to use this function, which is a bit confusing. I suggest we change this function to 
```suggestion
    s_byte_buf_init_from_file(
    struct aws_byte_buf *out_buf,
    struct aws_allocator *alloc,
    const char *filename,
    size_t size_hint)
```
and let `aws_byte_buf_init_from_file` calculate size_hint.

Reply:
I tried writing it that way first, but it was more confusing IMHO.

You ended up with 3 functions that needed to deal with error-handling instead of just 1, and scrambled logic around instead of putting it all in 1 place.